glib/src/serde.rs

```diff
@@ -0,0 +1,248 @@
+            }
+        }
+    };
+    ($ty:ty, String($arg:ident) => $body:block) => {
+        impl<'de> ::serde::Deserialize<'de> for $ty {
+            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+            where
+                D: ::serde::Deserializer<'de>,
+            {
+                struct Visitor;
+
+                impl<'a> ::serde::de::Visitor<'a> for Visitor {
+                    type Value = $ty;
+
+                    fn expecting(
+                        &self,
+                        formatter: &mut ::std::fmt::Formatter,
+                    ) -> ::std::fmt::Result {
+                        formatter.write_str("a valid UTF-8 string")
+                    }
+
+                    fn visit_str<E>(self, _: &str) -> Result<Self::Value, E>
+                    where
+                        E: de::Error,
+                    {
+                        todo!()
+                    }
+
+                    fn visit_string<E>(self, $arg: String) -> Result<Self::Value, E>
+                    where
+                        E: ::serde::de::Error,
+                    $body
+                }
+
+                deserializer.deserialize_string(Visitor)
+            }
+        }
+    };
+
+    ($ty:ident$(<$($generic:ident$(: $bound:tt $(+ $bound2:tt)*)?),+>)?, $expect:literal, Collection($arg:ident) => $body:block) => {
+        impl<'de: 'a, 'a: 'de, $($($generic $(: ::serde::Deserialize<'de> + $bound $(+ $bound2)*)?),+)?> ::serde::Deserialize<'de> for $ty$(<$($generic),+>)? {
+            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+            where
+                D: ::serde::Deserializer<'de>,
+            {
+                struct Visitor<'v, $($($generic $(: ::serde::Deserialize<'v> + $bound $(+ $bound2)*)?),+)?>(::std::marker::PhantomData<&'v ()>, $($(::std::marker::PhantomData<fn() -> $generic>),+)?);
+
+                impl<'a, $($($generic $(: ::serde::Deserialize<'a> + $bound $(+ $bound2)*)?),+)?> ::serde::de::Visitor<'a> for Visitor<'a, $($($generic),+)?> {
+                    type Value = $ty$(<$($generic),+>)?;
+
+                    fn expecting(&self, formatter: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+                        formatter.write_str($expect)
+                    }
+
+                    fn visit_seq<A>(self, mut $arg: A) -> Result<Self::Value, A::Error>
+                    where
+                        A: ::serde::de::SeqAccess<'a>,
+                    $body
+                }
+
+                deserializer.deserialize_seq(Visitor(::std::marker::PhantomData, $($(::std::marker::PhantomData::<fn() -> $generic>::default()),+)?))
+            }
+        }
+    };
+}
+
+mod byte_array {
+    use crate::ByteArray;
+
+    serialize_impl!(ByteArray, Bytes(b) => b);
+}
+
+mod bytes {
+    use crate::Bytes;
+
+    serialize_impl!(Bytes, Bytes(b) => b);
+
+    deserialize_impl!(
+        Bytes,
+        Bytes(seq) => {
+            let mut byte_vec = vec![];
```

Comment:
You could use the size hint here to reserve initial capacity

Reply:
Would this otherwise be acceptable?

```rs
let mut byte_vec = match seq.size_hint() {
	Some(size) => Vec::with_capacity(size),
	None => vec![],
};
```

Reply:
I think so. I would do the same as the deserialize impl for `Vec` is doing inside serde.

Reply:
Serde initializes `Vec<_>` with `Vec::with_capacity(std::cmp::min(size_hint.unwrap_or(0), 4096))`. Should this be done as well to `Slice<_>` and `PtrSlice<_>`? They can be instantiated with an initial capacity as well.